src/System.Text.Json/src/System/Text/Json/Serialization/ReadStackFrame.cs

```diff
@@ -202,9 +202,8 @@ public static object CreateEnumerableValue(ref ReadStack state)
 
                 state.Current.TempEnumerableValues = converterList;
 
-                // Clear the value if present to ensure we don't confuse tempEnumerableValues with the collection.
-                if (!jsonPropertyInfo.IsPropertyPolicy &&
-                    !state.Current.JsonPropertyInfo.IsImmutableArray)
+                // Clear the value if present to ensure we don't confuse TempEnumerableValues with the collection.
+                if (!jsonPropertyInfo.IsPropertyPolicy && jsonPropertyInfo.CanBeNull)
```

Comment:
We shouldn't attempt to set **any** value-type enumerable to `null`, not just `ImmutableArray`. Same below.

Given the intention to set the collection to null, there might be an issue with this check since value-type collections are excluded and this isn't accounted for anywhere. I'll dig into this as part of https://github.com/dotnet/corefx/issues/42399.

Reply:
I assume we want to throw when we see a `null` literal in a JSON array and the corresponding CLR element type is a value type. Wouldn't we just throw JsonException if `CanBeNull=true` here?

Reply:
> I assume we want to throw when we see a `null` literal in a JSON array and the corresponding CLR element type is a value type.

Yes, that's what we should (and currently) do.

In this area of the code, we are attempting to clear the collection property by setting it to null, so that we don't get confused in our null checks to `state.Current.Temp[Dictionary|Enumerable]Values` (are we using a collection converter?) later on. The `jsonPropertyInfo.CanBeNull` clause of this check guards against attempting to set a value type to null which I believe will give an `InvalidCastException`.
This means that we don't clear value type collections (which seem edge-casey). With the inital comment, I was noting that this _may_ be a bug since the intention was to clear all collections, and that I plan to look into this in a future PR for https://github.com/dotnet/corefx/issues/42399.

The issue for https://github.com/dotnet/corefx/issues/42399 is that we should throw `JsonException` when we see `null` where we expect a JSON array for a value type collection ([which we don't currently](https://github.com/dotnet/corefx/issues/42399)), e.g. [here](https://github.com/dotnet/corefx/blob/961bb667a056bb3b9a5f32e22aed8bb64eaddcad/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleNull.cs#L93-L94).

Reply:
> corresponding CLR element type is a value type.

I would imagine this shouldn't be true if it's a nullable value type (int?, DateTime?, etc.).

Reply:
> What's the behavior for that case?

[Those work as expected](https://dotnetfiddle.net/14qZfX). We execute tests for this via: https://github.com/dotnet/corefx/blob/ac02476cfd31e0aed74e5ea4c9dd0af3d7b1ef7f/src/System.Text.Json/tests/Serialization/TestClasses.SimpleTestClassWithNullables.cs#L10

Reply:
I was thinking about nullable value types as the T of the collections like immutable array, regular array. So `DateTime?[]` or `ImmutableArray<DateTime?>. Do these work as expected too?

As an aside, looks like nullable ImmutableArray breaks (https://dotnetfiddle.net/14qZfX):
```C#
Console.WriteLine(JsonSerializer.Deserialize<ImmutableArray<DateTime>?>("null") == null);
```
```text
Unhandled exception. System.ArgumentNullException: Value cannot be null.
   at System.RuntimeType.MakeGenericType(Type[] instantiation)
```

Reply:
> I was thinking about nullable value types as the T of the collections like immutable array, regular array. So `DateTime?[]` or `ImmutableArray<DateTime?>. Do these work as expected too?

I was thinking about nullable value types as the T of the collections like immutable array, regular array. So `DateTime?[]` or `ImmutableArray<DateTime?>. Do these work as expected too?

Yes we have tests for these:

https://github.com/dotnet/corefx/blob/6878693e4bdeb509abfdf51ad40a5a1aae46e2e4/src/System.Text.Json/tests/Serialization/NullableTests.cs#L15

https://github.com/dotnet/corefx/blob/6878693e4bdeb509abfdf51ad40a5a1aae46e2e4/src/System.Text.Json/tests/Serialization/NullableTests.cs#L198

> As an aside, looks like nullable ImmutableArray breaks (https://dotnetfiddle.net/14qZfX)

Yes, this is an extension of the known issue with assigning JSON null to value-type collections: https://github.com/dotnet/corefx/issues/42399#issuecomment-553190419.